### ISSUE TITLE
Auto Interface Implementation Differing Column Names, Parameter Names, Property Names

### ISSUE THREAD

**Issue by buist-sharris** [jonwagner/Insight.Database#417]
I am trying to implement Auto Interfaces for objects that have property names that don't fully line up with column names that don't fully line up with parameter names.  When I try to create an instance of the interface, I get an Invalid Operation Exception "Special Parameter transaction must have type System.Data.IDbTransaction".

**Table: MyTransactions**
Transactionamount decimal
TransactionDate datetime
TransactionDescription varchar
Employee_Code varchar

**Stored Proc:**
```
InsertTransaction (@transactionAmount decimal (18,2),@transactionDate datetime,@transactionDescription varchar(255),@employeeCode varchar(11))

INSERT INTO...
```

**Class: MyTransaction**

         [Column("TransactionDate")]
        public DateTime TransactionDate {get; set;}
		
        [Column("TransactionAmount")]
        public decimal Amount {get; set;}

        [Column("TransactionDescription")]
        public string Desription {get; set;}

        [Column("Employee_Code")]
        public string EmployeeNumber {get; set;}

        [RecordId]
        [Column("TransactionId")]
        public int? Id {get; set;}

**Auto Interface: ITransactionRepo**
`
void InsertTransaction (MyTransaction transaction);`

**Code that generates error:**
`
interface ITransactionRepo : IDbConnection, IDbTransaction
{
using (var repo = db.OpenAs<ITransactionRepo>())
}
`

I have tried adding a custom parameter mapper, but feel I'm probably not doing it correctly...
```
private static async Task Main(string[] args)
        {
            SqlInsightDbProvider.RegisterProvider();
            ColumnMapping.Parameters.AddMapper(new MyParameterMapper());
        }

public class MyParameterMapper : IParameterMapper
    {
        public string MapParameter(Type type, IDbCommand command, IDataParameter parameter)
        {
            if (command.CommandText != "InsertTransaction")
                return null; // null allows another handler to try

            if (parameter.ParameterName == "employeeCode")
                return "EmployeeNumber"; // a string value specifies the name of the field/property

            if (parameter.ParameterName == "transactionDescription")
                return "Desription"; // a string value specifies the name of the field/property

            if (parameter.ParameterName == "transactionAmount")
                return "Amount"; // a string value specifies the name of the field/property

            return "__unmapped__"; // returning an invalid string value prevents other handlers from mapping 
        }
    }
```

Is there any way to work with this to get everything to line up?

**Comment by jonwagner** [comment]
The issue is that Insight reserves the parameter name “transaction”. It’s for the ability to make multiple database calls participate in the same database transaction.

All of the other parameters are probably working fine. See if you can can get that one mapping to work.

On Dec 20, 2019, at 10:12 AM, buist-sharris <notifications@github.com> wrote:

﻿

I am trying to implement Auto Interfaces for objects that have property names that don't fully line up with column names that don't fully line up with parameter names. When I try to create an instance of the interface, I get an Invalid Operation Exception "Special Parameter transaction must have type System.Data.IDbTransaction".

Table: MyTransactions
Transactionamount decimal
TransactionDate datetime
TransactionDescription varchar
Employee_Code varchar

Stored Proc: InsertTransaction (@transactionAmount decimal (18,2),@transactionDate datetime,,@transactionDescription varchar(255),@employeeCode varchar(11))

Class: MyTransaction
[Column("TransactionDate")]
public DateTime TransactionDate {get; set;}

    [Column("TransactionAmount")]
    public decimal Amount {get; set;}

    [Column("TransactionDescription")]
    public string Desription {get; set;}

    [Column("Employee_Code")]
    public string EmployeeNumber {get; set;}

    [RecordId]
    [Column("TransactionId")]
    public int? Id {get; set;}


Auto Interface: ITransactionRepo
void InsertTransaction (MyTransaction transaction);

Code that generates error:
using (var repo = db.OpenAs())

I have tried adding a custom parameter mapper, but feel I'm probably not doing it correctly...
private static async Task Main(string[] args)
{
SqlInsightDbProvider.RegisterProvider();
ColumnMapping.Parameters.AddMapper(new MyParameterMapper());
}

public class MyParameterMapper : IParameterMapper
{
public string MapParameter(Type type, IDbCommand command, IDataParameter parameter)
{
if (command.CommandText != "InsertTransaction")
return null; // null allows another handler to try

        if (parameter.ParameterName == "employeeCode")
            return "EmployeeNumber"; // a string value specifies the name of the field/property

        if (parameter.ParameterName == "transactionDescription")
            return "Desription"; // a string value specifies the name of the field/property

        if (parameter.ParameterName == "transactionAmount")
            return "Amount"; // a string value specifies the name of the field/property

        return "__unmapped__"; // returning an invalid string value prevents other handlers from mapping
    }
}


Is there any way to work with this to get everything to line up?

**Comment by buist-sharris** [comment]
Changing the argument name from "transaction" resolved the initial error when trying to create an instance of the auto interface.  However I now get an object null exception being thrown by the ParameterMapper.  When I step into it, the first iteration through when my interface insert method is executed is for "RETURN_VALUE", doesn't match any of my mapping check, and then return "__unmapped__".  When it returns to the code, the object null exception is thrown.

**Comment by buist-sharris** [comment]
I solved it.  Returning null instead of an invalid value...

`public string MapParameter(Type type, IDbCommand command, IDataParameter parameter)
        {
            if (command.CommandText == "dbo.InsertTransaction")
            {
                if (parameter.ParameterName == "employeeCode")
                    return "EmployeeNumber";

                if (parameter.ParameterName == "transactionDescription")
                    return "Desription"; 

                if (parameter.ParameterName == "transactionAmount")
                    return "Amount";
            }

            return null;
        }`

**Comment by jonwagner** [comment]
And with the bug fix for #423 you should be good to go.
